pkg/kubelet/container/hook/hook.go

```diff
@@ -0,0 +1,164 @@
+// have meaning in response to the PRE_CREATE event.
+//
+// {
+//   "pod": { ... },
+//   "container": { ... },
+//   "config": { ... }
+// }
+//
+// Finally, bytes the hook program outputs to standard error will be included
+// in the Kubelet logs.
+func (h scriptContainerEventHook) callWithEvent(
+	eType eventType,
+	pod *api.Pod,
+	container *api.Container,
+	config *dockertypes.ContainerCreateConfig) (*eventContext, error) {
+
+	e := eventInfo{eType, eventContext{pod, container, config}}
+	data, err := json.Marshal(e)
+	if err != nil {
+		return nil, err
+	}
+	cmd := exec.Command(h.scriptPath)
+	cmd.Stdin = strings.NewReader(string(data))
+	cmd.Env = []string{} // Clean subprocess environment.
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, err
+	}
+	cmd.Start()
+
+	// Read standard error and log it, in a separate goroutine
+	// to avoid blocking writes to the stdout pipe.
+	go func() {
+		eout, err := ioutil.ReadAll(stderr)
+		if err != nil {
+			glog.Errorf("Unable to read stderr from hook script [%s]: %s", h.scriptPath, err.Error())
+		}
+		glog.Infof("Container event hook script [stderr]:\n%s", string(eout))
+	}()
+
+	// Parse standard out as eventContext JSON.
+	var resultContext *eventContext
+	if err := json.NewDecoder(stdout).Decode(&resultContext); err != nil {
+		return nil, err
+	}
+
+	// Log returned context.
+	resultJSON, err := json.Marshal(resultContext)
+	if err != nil {
+		return nil, err
+	}
+	glog.Infof("Container event hook script (%s) returned pod, container and config:\n%s", eType, resultJSON)
+
+	return resultContext, nil
+}
+
+func (h scriptContainerEventHook) Name() string {
+	return fmt.Sprintf("Script container event hook [%s]", h.scriptPath)
+}
+
+func (h scriptContainerEventHook) PreCreate(pod *api.Pod, container *api.Container, config *dockertypes.ContainerCreateConfig) {
+	ctx, err := h.callWithEvent(preCreate, pod, container, config)
+	if err != nil {
+		glog.Errorf("Failed to execute hook script [%s] for event %s: %s", h.scriptPath, preCreate, err.Error())
+		return
+	}
+
+	// Modify referenced pod, container and config.
+	*pod = *ctx.Pod
+	*container = *ctx.Container
+	*config = *ctx.Config
+}
+
+func (h scriptContainerEventHook) PreStart(pod *api.Pod, container *api.Container, config *dockertypes.ContainerCreateConfig) {
+	_, err := h.callWithEvent(preStart, pod, container, config)
+	if err != nil {
+		glog.Errorf("Failed to execute hook script [%s] for event %s: %s", h.scriptPath, preStart, err.Error())
+		return
+	}
+}
+
+func (h scriptContainerEventHook) PostStart(pod *api.Pod, container *api.Container, config *dockertypes.ContainerCreateConfig) {
+	_, err := h.callWithEvent(postStart, pod, container, config)
+	if err != nil {
+		glog.Errorf("Failed to execute hook script [%s] for event %s: %s", h.scriptPath, postStart, err.Error())
+		return
+	}
```

Comment:
Is there any scenario where we need to modify the pod, container or the docker config post start?